storage/src/backend/localdisk.rs

```diff
@@ -0,0 +1,246 @@
+            error!("device path is required");
+            return Err(einval!("device path is required"));
+        }
+
+        let path_buf = Path::new(&path).to_path_buf().canonicalize().map_err(|e| {
+            error!("failed to parse path {}: {:?}", path, e);
+            e
+        })?;
+
+        let file = OpenOptions::new().read(true).open(&path_buf).map_err(|e| {
+            error!("failed to open {}: {:?}", path, e);
+            e
+        })?;
+
+        let local_disk = LocalDisk {
+            device_file: file,
+            device_path: path,
+            metrics: BackendMetrics::new(id, "localdisk"),
+            entries: RwLock::new(HashMap::new()),
+        };
+
+        local_disk.init().map_err(|e| {
+            error!("load disk failed {}: {:?}", local_disk.device_path, e);
+            e
+        })?;
+
+        Ok(local_disk)
+    }
+
+    pub fn init(&self) -> Result<()> {
+        let mut table_guard = self.entries.write().unwrap();
+
+        // Open disk image.
+        let cfg = gpt::GptConfig::new().writable(false);
+        let disk = cfg.open(self.device_path.clone())?;
+        let partitions = disk.partitions();
+        let sector_size = gpt::disk::DEFAULT_SECTOR_SIZE;
+        info!(
+            "Localdisk initialized at {}, has {} patitions, GUID: {}",
+            self.device_path,
+            partitions.len(),
+            disk.guid()
+        );
+
+        for (k, v) in partitions {
+            let length = v.bytes_len(sector_size)?;
+            let base_offset = v.bytes_start(sector_size)?;
+            if base_offset.checked_add(length).is_none() {
+                error!("partition {} ends with an invalid offset", v.part_guid);
+                return Err(eio!("partition ends with an invalid offset"));
+            };
+            let last_offset = base_offset + length;
+            let name = v.name.clone();
+
+            if name.is_empty() {
+                error!("partition {} does not record an blob id", v.part_guid);
+                return Err(eio!("partition does not record an blob id"));
+            }
+
+            let partition = Arc::new(LocalDiskPartition {
+                blob_id: name.clone(),
+                device_fd: self.device_file.as_raw_fd(),
+                base_offset,
+                last_offset,
+                length,
+                metrics: self.metrics.clone(),
+            });
+
+            debug!(
+                "Localdisk partition {} initialized, blob id: {}, offset from {} to {}, length {}",
+                k,
+                partition.blob_id,
+                partition.base_offset,
+                partition.last_offset,
+                partition.length
+            );
+            table_guard.insert(name, partition);
+        }
+
+        Ok(())
+    }
+
+    // Disk names in GPT tables cannot store full 64-byte blob ids, so we should truncate them to 32 bytes.
+    fn truncate_blob_id(blob_id: &str) -> LocalDiskResult<&str> {
```

Comment:
How about storing binary rather than strings? Then 32 bytes can hold the whole blob id

Reply:
> How about storing binary rather than strings? Then 32 bytes can hold the whole blob id

Very good suggestion, I will try to modify this.

Reply:
I tried to use the storage method of Hex to String, but there are still problems in the implementation, and I will improve it later.

In the future, I will use two different methods to obtain the blob_id in get_blob() to maintain compatibility with different versions of the localdisk image.